### PR TITLE
METAL-3262 proxy serviceMonitor monitoring

### PR DESCRIPTION
**Pull request by dabde**
Adding possibility to create a related service monitor resource for the prometheus operator, when creating a proxy for a db.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -47,6 +48,8 @@ func init() {
 
 	utilruntime.Must(kcirocksv1alpha1.AddToScheme(scheme))
 	//+kubebuilder:scaffold:scheme
+
+	thirdpartyapi.AppendToScheme(scheme)
```
Comment (dabde) [inline main.go:52]: don't know where is the right place to extend the scheme with thridparty api's.

File: pkg/utils/proxy/cloudproxy.go [hunk 2]
```diff
@@ -174,3 +175,26 @@ func (cp *CloudProxy) container() (v1.Container, error) {
...
+			Endpoints: []promv1.Endpoint{Endpoint},
+			NamespaceSelector: promv1.NamespaceSelector{
+				MatchNames: []string{cp.Namespace},
+			},
+			Selector: metav1.LabelSelector{
+				MatchLabels: cp.Labels,
+			},
+		},
+	}, nil
+}
```
Comment (dabde) [inline pkg/utils/proxy/cloudproxy.go:200]: will generate and create something like this:
```yaml
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  creationTimestamp: "2022-02-03T12:50:22Z"
  generation: 1
  labels:
    app: cloudproxy
    created-by: db-operator
    db-name: example-db
  name: db-example-db-sm
  namespace: operator
  resourceVersion: "9039"
  uid: eb4a365a-de53-4a50-b41a-712be83b240e
spec:
  endpoints:
  - bearerTokenSecret:
      key: ""
    port: metrics
  namespaceSelector:
    matchNames:
    - operator
  selector:
    matchLabels:
      app: cloudproxy
      created-by: db-operator
      db-name: example-db
```

File: pkg/utils/proxy/cloudproxy.go [hunk 3]
```diff
@@ -174,3 +175,26 @@ func (cp *CloudProxy) container() (v1.Container, error) {
 func (cp *CloudProxy) buildConfigMap() (*v1.ConfigMap, error) {
 	return nil, nil
 }
+
+func (cp *CloudProxy) buildPromService() (*promv1.ServiceMonitor, error) {
```
Comment (hyunysmile) [inline pkg/utils/proxy/cloudproxy.go:179]: `buildServiceMonitor()` is more straightforward

File: pkg/utils/proxy/create.go [hunk 4]
```diff
@@ -54,3 +55,14 @@ func BuildConfigmap(proxy Proxy) (*v1.ConfigMap, error) {
 
 	return cm, nil
 }
+
+// BuildPromService builds kubernetes prometheus ServiceMonitor CR object used for monitoring
+func BuildPromService(proxy Proxy) (*promv1.ServiceMonitor, error) {
```
Comment (hyunysmile) [inline pkg/utils/proxy/create.go:60]: also here I would name `BuildServiceMonitor` like other functions. 
including resource name as it is, matchs with current convention.

File: controllers/helper.go [hunk 5]
```diff
@@ -78,3 +79,12 @@ func containsString(slice []string, s string) bool {
 	}
 	return false
 }
+
+func checkCRD(crds crdv1.CustomResourceDefinitionList, api string) bool {
```
Comment (hyunysmile) [inline controllers/helper.go:83]: what about `isCrdInstalled`

**Comment by hyunysmile** [comment]
can you clean this part? https://github.com/kloeckner-i/db-operator/blob/master/pkg/utils/proxy/cloudproxy.go#L111

File: controllers/database_controller.go [hunk 6]
```diff
@@ -501,6 +502,35 @@ func (r *DatabaseReconciler) createProxy(ctx context.Context, dbcr *kciv1alpha1.
 		}
 	}
 
+	crdList := crdv1.CustomResourceDefinitionList{}
```
Comment (hyunysmile) [inline controllers/database_controller.go:505]: can you please add `if dbcr.IsMonitoringEnabled()`

Comment (dabde) [reply controllers/database_controller.go:505]: added over the proxyinterface.

File: pkg/utils/proxy/cloudproxy.go (hunk 3 above)
Comment (dabde) [reply pkg/utils/proxy/cloudproxy.go]: changed.

File: pkg/utils/proxy/create.go (hunk 4 above)
Comment (dabde) [reply pkg/utils/proxy/create.go]: changed.

File: controllers/helper.go (hunk 5 above)
Comment (dabde) [reply controllers/helper.go]: change to `inCrdList()`, think `isCrdInstalled` is not correct, related the function is only compared an already fetched list. Was think to rename to `isCrdInstalled`, but this means to reference the `DatabaseReconciler` outside of the `database_controller.go`. Don't know if we want to limit on this, but was not feel right for me.

**Review by hyunysmile — APPROVED** [review]

**Pull request merged**